src/main/java/io/quarkus/gizmo2/impl/EqualsHashCodeToStringGenerator.java

```diff
@@ -0,0 +1,297 @@
+
+public class EqualsHashCodeToStringGenerator {
+    private final ClassCreator cc;
+    private final List<FieldDesc> fields;
+
+    public EqualsHashCodeToStringGenerator(ClassCreator cc, List<FieldDesc> fields) {
+        this.cc = cc;
+        this.fields = fields;
+    }
+
+    public void generateEquals() {
+        MethodDesc floatToIntBits = MethodDesc.of(Float.class, "floatToIntBits", int.class, float.class);
+        MethodDesc doubleToLongBits = MethodDesc.of(Double.class, "doubleToLongBits", long.class, double.class);
+
+        MethodDesc booleanArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, boolean[].class, boolean[].class);
+        MethodDesc byteArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, byte[].class, byte[].class);
+        MethodDesc shortArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, short[].class, short[].class);
+        MethodDesc intArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, int[].class, int[].class);
+        MethodDesc longArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, long[].class, long[].class);
+        MethodDesc floatArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, float[].class, float[].class);
+        MethodDesc doubleArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, double[].class, double[].class);
+        MethodDesc charArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, char[].class, char[].class);
+        MethodDesc objectEquals = MethodDesc.of(Objects.class, "equals", boolean.class, Object.class, Object.class);
+        MethodDesc objectArrayEquals = MethodDesc.of(Arrays.class, "equals", boolean.class, Object[].class, Object[].class);
+        MethodDesc arrayDeepEquals = MethodDesc.of(Arrays.class, "deepEquals", boolean.class, Object[].class, Object[].class);
+
+        ClassDesc thisClass = cc.type();
+
+        cc.method("equals", mc -> {
+            mc.public_();
+            mc.returning(boolean.class);
+            ParamVar other = mc.parameter("other", Object.class);
+            mc.body(b0 -> {
+                b0.if_(b0.eq(mc.this_(), other), b1 -> {
+                    b1.return_(true);
+                });
+                b0.ifNotInstanceOf(other, thisClass, b1 -> {
+                    b1.return_(false);
+                });
+
+                Expr otherCast = b0.define("other", b0.cast(other, thisClass));
+                for (FieldDesc field : fields) {
+                    if (!field.owner().equals(thisClass)) {
+                        throw new IllegalArgumentException("Field does not belong to " + thisClass.displayName() + ": " + field);
+                    }
+
+                    Expr thisValue = b0.define("thisValue", b0.get(mc.this_().field(field)));
+                    Expr thatValue = b0.define("thatValue", b0.get(otherCast.field(field)));
+                    switch (field.type().descriptorString()) {
```

Comment:
These repeated cases where it's a `switch` of a descriptor string, but then also an `if` tree, might be better written (sometimes at least) as `switch (descriptorString.charAt(0)`, with a sub-switch for the `[` case. This would eliminate the extra `if` tree and might end up looking marginally cleaner, maybe?